### PR TITLE
Separate projection from query execution

### LINKED ISSUE #1120: Separate projection from query execution
e.g.

```java
query.from(customer).where(customer.firstName.eq("Tom")).list(customer)
```

into

```java
query.select(customer).from(customer).where(customer.firstName.eq("Tom")).list()
```

This has the following benefits

* queries can be written both with select definition in the beginning as well as the end
* subqueries can be defined via normal Query instances
* Projectable and SimpleProjectable can be merged into the same class

Disadvantages
* more verbose syntax in some cases
* generic parameter in Query interface/class

### PR DESCRIPTION
**Pull request by timowest**
Fixes #1120

doesn't have a backport, since it is a backwards incompatible change

Does not yet compile and also SQL part has not yet been refactored, major SQL related PRs should be merged before this is continued.

### CODE REVIEW

**Comment by timowest** [comment]
This PR contains now the first steps of the refactoring to the new query model. Although some usages become now a bit more verbose, the new model feels much more flexible.

**Comment by timowest** [comment]
I am not yet quite happy with ProjectableQuery, maybe Selectable or something might be better?

**Comment by timowest** [comment]
@Shredder121 @johnktims  Finally building now. Please review ;)

Some other changes done 
* rename from Projectable to Fetchable
* QueryMetadata validation off by default
* experimental Scala query types removed

**Comment by timowest** [comment]
Let's keep this in progress until we finalize the method names for `Fetchable`

File: querydsl-collections/src/main/java/com/querydsl/collections/CollDeleteClause.java [hunk 1]
```diff
@@ -33,10 +33,10 @@
 
     private final Path<T> expr;
 
-    private final CollQuery query;
+    private final CollQuery<T> query;
 
     public CollDeleteClause(QueryEngine qe, Path<T> expr, Collection<? extends T> col) {
-        this.query = new CollQuery(qe).from(expr, col);
+        this.query = new CollQuery<Void>(qe).from(expr, col).select(expr);
```
Comment (Shredder121) [inline querydsl-collections/src/main/java/com/querydsl/collections/CollDeleteClause.java]: Maybe in favor of `new XXXQuery<Void>` we can add static factory methods that return a `XXXQuery<?>`.
Essentially to hide the not yet known parameter.
I recall having discussed this, but I think it's good to recap.

File: querydsl-collections/src/main/java/com/querydsl/collections/CollQuery.java [hunk 2]
```diff
@@ -74,16 +73,19 @@ public CollQuery(QueryMetadata metadata, QueryEngine queryEngine) {
...
 
-    /**
-     * @return
-     */
     @Override
-    public QueryMetadata getMetadata() {
-        return queryMixin.getMetadata();
+    public <E> CollQuery<E> select(Expression<E> expr) {
+        queryMixin.setProjection(expr);
+        return (CollQuery<E>)this;
```
Comment (Shredder121) [inline querydsl-collections/src/main/java/com/querydsl/collections/CollQuery.java]: @timowest have you by chance experimented with letting the querymixin hold the result type? Is that an option?

File: querydsl-collections/src/main/java/com/querydsl/collections/CollDeleteClause.java (hunk 1 above)
Comment (timowest) [inline querydsl-collections/src/main/java/com/querydsl/collections/CollDeleteClause.java]: Good point. I will do that.

File: querydsl-collections/src/main/java/com/querydsl/collections/CollQuery.java (hunk 2 above)
Comment (timowest) [inline querydsl-collections/src/main/java/com/querydsl/collections/CollQuery.java]: I am not sure. QueryMixin has the query type as an argument, but the select methods change it, so I guess there will be anyway a cast.

**Comment by timowest** [comment]
Concerning finishing the javadocs, I'd like to continue that on a per module basis, that's why there are some empty stubs.
So in order to continue with the javadoc fixes, it would be good to get this merged.

**Comment by Shredder121** [comment]
I see, I do agree with that.

**Pull request merged**